### ISSUE TITLE
Introduce Analytics account provisioning JS datastore actions/selectors and component

### ISSUE THREAD

**Issue by adamsilverstein** [google/site-kit-wp#1211]
## Feature Description

Show a helpful screen to users who don't have an Analytics account, they get a setup screen with a "Setup Account" action button. (Note: the exact UI/language for this screen is still to be determined.)

### Details
When activating Analytics, the user first goes thru the auth flow, then arrives at the Analytics setup screen. When no user account is detected, the view should show a (new) screen with all the [required GA account fields](https://developers.google.com/analytics/devguides/config/provisioning/v3/user-controlled#technical-ticket): eg. account name, website url, profile name, view name and an action button. 

Clicking the action button sends all fields to a REST request to the plugin (to a new `provision` data endpoint covered in a separate issue). 

Clicking the action button also sends an analytics event tracking event:
`trackEvent( 'analytics_setup', 'new_account_setup_clicked' );`

Display loading indicator on submission, then handle success or error response.

Error screen: Language/UI TBD.

Success Screen:
* Language (TBD) to confirm the account was successfully created “Configure Analytics” button takes the user to the dashboard.
* Clicking button also sends tracking event:
`trackEvent( 'analytics_setup', 'analytics_configured' );`

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

Users with no GA account or users who choose to create a new account should be able to do so directly from the plugin. 
Full description of the flow is already in "Details" above, high-level ACs here: 
* Display new screen with all necessary fields (as many as possible pre-filled) for the user to approve or edit
    * Account: pre-filled name (Site Title) --> user should be able to edit this 
    * Property: pre-filled domain of the site URL --> user should be able to edit this 
    * Profile: pre-filled (all website data) --> user should be able to edit this 
    * Timezone: pre-filled (from WP site's timezone) --> user should be able to pick this from dropdown 
* Redirect smoothly to GA TOS for the user to approve (see https://developers.google.com/analytics/devguides/config/provisioning/v3/user-controlled#technical-tos-flow-url), using the account ticket ID generated via #1212
* Redirect to main plugin dashboard and show success message for the new account + it will take some time for data to aggregate (top banner) 
* Handle errors during the process and display relevant messages for user to take action 
* Finally take the user to the main dashboard with the standard congrats banner at the top.

## Implementation Brief

* Create a new component to display when the user has no Analytics account: `AnalyticsSetupNoAccountNotice`.
* Integrate the component into the setup flow when the no account is detected during the initial check in: `assets/js/modules/analytics/common/index.js` - use the new component instead of `account-create-legacy`
* The `AnalyticsSetupNoAccountNotice` component includes:
  - Intro text (exact language TBD): 'Create a new Analytics account"
  - A form containing text fields for (and with the id) `accountName`, `propertyName, `profileName` and `timezone`.
  - Form fields are pre-filled and user editable: 
    - `accountName` should default to the site title.
    - `propertyName` should default to the site's domain/host.
    - `profileName` should default to "All website data".
  - The `timezone` field is a dropdown of timezones, with the WordPress site timezone preselected. The fields in this dropdown should match what the user sees on the WordPress settings screen. To construct the dropdown, we can follow the login in `wp_timezone_choice` which builds the wp settings dropdown: https://github.com/WordPress/wordpress-develop/blob/c8d918e47634bd423e0c7ccfc0e8f7a34515c3f5/src/wp-includes/functions.php#L5714. (Note: I considered using a 3rd party Timezone dropdown component, eg. https://www.npmjs.com/package/select-timezone-material-ui, however this adds a ton of code for a simple dropdown. We already have a dropdown selector, so we only need to provide the correct list. WordPress already provides a model for how to do that, and using the same list as the settings screen will be the best experience for users.)
  - A "Create account" button at the form will be disabled unless all fields are complete (eg if the user clears a field, the button becomes disabled)
  - When the user clicks the button, send an analytics event: `trackEvent( 'analytics_setup', 'new_account_setup_clicked' );`
* Collect the form data and submit using the datastore:
 - In `assets/js/modules/analytics/datastore/accounts.js`:
   - A `createAccount` action that takes an object with `accountName, propertyName, profileName, timezone` and yields an action object with a type `FETCH_CREATE_ACCOUNT`.
   - Handlers for `receiveCreateAccount` and `receiveCreateAccountFailed` to handle the action response.
   - A new `FETCH_CREATE_ACCOUNT` section in `controls` that sends the payload to API.get: `return API.get( 'modules', 'analytics', 'create-account-ticket', { accountName, propertyName, profileName, timezone, } );`
   - A `receiveCreateAccount` action that redirects the user to accept the Google Analytics [Terms of Service] https://analytics.google.com/analytics/web/?provisioningSignup=false#management/TermsOfService/?api.accountTicketId={account_ticket_id} using the returned account_ticket->id.
   - A `RECEIVE_CREATE_ACCOUNT_FAILED` reducer to place the error into the store.
* Display any error response and possible remedy: (https://developers.google.com/analytics/devguides/config/provisioning/v3/errors) 

* The user will travel thru the provisioning process, [completing the Terms of Service](https://github.com/google/site-kit-wp/issues/1213). Upon success they are redirected back to the dashboard with a confirmation banner. On error they are redirected back to the setup screen with an error code.

* If the setup screen loads with an `error` code query variable, the setup page should display an error message (match existing error messaging) - depending on the error code:
  - `user_cancel` - tell the user that they must agree to the terms of service to complete the process
  - `max_accounts_reached` - Tell the user they have reached their max accounts. Seems unlikely we would ever see this since we only use the provisioning request for users with no accounts, however it is possible the user could switch to another Google account during the auth flow and get this error returned.
  - `backend_error` - unknown error: tell the user to wait a few minutes and try again.
  

## QA Brief

* 

## Changelog entry

* N/A

**Comment by marrrmarrr** [comment]
@adamsilverstein is it feasible here to skip the "configure Analytics" step to save the user a click? Can we do that on their behalf?

**Comment by adamsilverstein** [comment]
>  is it feasible here to skip the "configure Analytics" step to save the user a click? Can we do that on their behalf?

@marrrmarrr -

Yes we certainly can with a redirect to the dashboard when everything succeeds.

My concern was that if we drop them right into the dashboard, they might not understand that the account setup worked. I guess we do show a success message on the Dashboard already "Congrats on setting up Analytics" (I guess I've become a little blind to this notice though). So technically possible, best UX up to your discretion.

**Comment by felixarntz** [comment]
@adamsilverstein This is blocked by #1101 and at least deeply intertwined with that. Let's work on this one early next week, and check in later this week whether we can already do an IB based on where #1101 is at that point.

**Comment by adamsilverstein** [comment]
@felixarntz Even though this task is blocked in part by the refactoring work, I took a pass at the IB for the parts we know we need to build, and how. The main points missing are: how to integrate the new component into the setup flow, and how to send the data POST when the user clicks the button.

**Comment by felixarntz** [comment]
IB ✅ , with some quirks to consider:
* The copy must also work for when the user _chose_ to create an account. So it shouldn't say "We didn't find an account" or similar.
* The `propertyName` field should default to the site's domain/host, not the full URL.
* We need to ensure that the WordPress timezones are part of the TZ database (see https://developers.google.com/analytics/devguides/config/mgmt/v3/mgmtReference/management/profiles#timezone). I suppose they are, but there may be some weird WordPress-specific ones. Let's check on a high level for the different types of timezone formats.
* `RECEIVE_CREATE_ACCOUNT_ERROR` --> `RECEIVE_CREATE_ACCOUNT_FAILED`
* I'm not sure we need a `getCreateAccountTicket` selector. I think we can redirect from within the datastore's `createAccount` action, it should just be documented in the action doc block. The redirect is technically a required part of creating an account, so it makes sense to include it there.

**Comment by adamsilverstein** [comment]
> The copy must also work for when the user chose to create an account. So it shouldn't say "We didn't find an account" or similar.

Updated header to read: "Create a new Analytics account"

> The propertyName field should default to the site's domain/host, not the full URL.

Adjusted and cleaned up that section.

> We need to ensure that the WordPress timezones are part of the TZ database (see developers.google.com/analytics/devguides/config/mgmt/v3/mgmtReference/management/profiles#timezone). I suppose they are, but there may be some weird WordPress-specific ones. Let's check on a high level for the different types of timezone formats.

I could not find a specific reference on the Analytics documentation site about what timezones are supported, this may be it (I found a reference to it in other Google APIs): https://github.com/unicode-org/cldr/blob/master/common/bcp47/timezone.xml. 

I think the WordPress list is a bit longer: if we use this list, then in cases where the selected WordPress timezone cannot be found, we can make the user pick a timezone from the list.

> RECEIVE_CREATE_ACCOUNT_ERROR --> RECEIVE_CREATE_ACCOUNT_FAILED

Fixed

> I'm not sure we need a getCreateAccountTicket selector. I think we can redirect from within the datastore's createAccount action, it should just be documented in the action doc block. The redirect is technically a required part of creating an account, so it makes sense to include it there.

Ok, adjusting to move the logic there.

**Comment by cole10up** [comment]
**Tested**

Built latest Site Kit zip. Installed and activated. Installed the tester plugin and flipped to PR: 1442. Setup SK and activated the Analytics module with a gmail account that doesn't have an Analytics account:

### **Proxy testing**
During proxy notice:
![image](https://user-images.githubusercontent.com/57115907/81891851-660fd700-955e-11ea-8716-69184d2b74f7.png)

Clicked Cancel during the screen below:
![image](https://user-images.githubusercontent.com/57115907/81891879-7b850100-955e-11ea-910e-14ea33f1789c.png)

Notice user is sent back to WP dashboard with a few errors:
![image](https://user-images.githubusercontent.com/57115907/81892145-32817c80-955f-11ea-8870-14d73f051d83.png)

Retrying from Dashboard reauth link I approved the proxy message below:
![image](https://user-images.githubusercontent.com/57115907/81891851-660fd700-955e-11ea-8716-69184d2b74f7.png)

Then I unchecked the box during this screen and clicked approve:
![image](https://user-images.githubusercontent.com/57115907/81891879-7b850100-955e-11ea-910e-14ea33f1789c.png)

Notice the proxy approved the setup:
![Sign in - Google Accounts](https://user-images.githubusercontent.com/57115907/81892311-a885e380-955f-11ea-9f09-e90384c707d9.gif)

@adamsilverstein - Is this expected?

I tried same scenario above with a different instance of Jurassic Ninja and am experiencing the 'Issue accessing data banner' and Create new Analytics account flow

![image](https://user-images.githubusercontent.com/57115907/81892744-b5ef9d80-9560-11ea-87b8-0741cf5e8b85.png)

Clicking Cancel on the create account forms page drives user back to the Proxy here:
![image](https://user-images.githubusercontent.com/57115907/81892968-3f9f6b00-9561-11ea-88a7-2dc3904b7e9e.png)


### **Account Creation testing fields**

**Observations**
- User is able to edit the 'View' field to gibberish and click 'Create Account' without an error message being displayed. The page just refreshes

- User is able to edit the 'Property' field adding in an invalid host URL and click 'Create Account' notice no error message displayed. Page just refreshes

**Happy Path**
Tried to complete the flow by filling out fields - notice page refreshes without on page errors. Console shows:

![image](https://user-images.githubusercontent.com/57115907/81893302-3ebb0900-9562-11ea-991e-59f028fa0312.png)

Tried another host, same issue

![image](https://user-images.githubusercontent.com/57115907/81893398-79bd3c80-9562-11ea-8f89-d20eaafb554c.png)

Sending back to CR to adjust

**Comment by aaemnnosttv** [comment]
@cole10up 
> * User is able to edit the 'View' field to gibberish and click 'Create Account' without an error message being displayed. The page just refreshes
>
> * User is able to edit the 'Property' field adding in an invalid host URL and click 'Create Account' notice no error message displayed. Page just refreshes

These fields are just for the _names_ of the entities that will be created for each, the only requirement currently is that they aren't empty.

Good catch about the ungranted scope!

**Comment by adamsilverstein** [comment]
> Notice the proxy approved the setup:
> @adamsilverstein - Is this expected?

I don't think so, we should detect that the user hasn't granted the full scopes and prevent them from trying to create an account. 

Where/how do we handle insufficient scopes elsewhere in the plugin? 

I will take a look.

**Comment by adamsilverstein** [comment]
> I tried same scenario above with a different instance of Jurassic Ninja and am experiencing the 'Issue accessing data banner' and Create new Analytics account flow

That is probably expected if you haven't granted the correct scopes. Maybe we should avoid showing the setup form entirely in this case?

I wonder what happens with the current version of the plugin if you enable analytics and don't grant all scopes during the flow? eg. is this a regression?

**Comment by adamsilverstein** [comment]
![image](https://user-images.githubusercontent.com/2676022/81943844-232f1d00-95b9-11ea-9bb9-409b7939d28d.png)

@cole This was caused by a bug introduced late in the development cycle that caused the provisioning request to go directly to the API vs going thru the proxy service. This is now corrected, can you try testing against https://github.com/google/site-kit-wp/pull/1343 again (using the tester plugin switch to develop, then back to the PR to re-install the latest build).

Also: I recommend activating the staging service plugin so you are using the staging proxy: the full set of changes there are not yet in production.

**Comment by adamsilverstein** [comment]
Regarding the scopes warning, the biggest issue here is we aren't displaying an error message for the request failure. I had noticed this before when my PHP fataled and I was getting an error response. I will try adding the <ErrorComponent> again or see how this scope error is displayed currently and match that.

A secondary issue is that we should be immediately detecting when a user fails to grant the correct scopes, eg. when they return from the Auth flow we should check the (returned) scopes immediately (I though we already did this?).

### Current version
I tested trying to create a profile with the **current release version** of the plugin without granting the management scope. This fails, with a helpful error message:

Skip management scope:
![image](https://user-images.githubusercontent.com/2676022/81948031-7d7eac80-95be-11ea-9e5d-17da340e8754.png)


Create new property:
![image](https://user-images.githubusercontent.com/2676022/81947733-24167d80-95be-11ea-821b-fe5462156faa.png)

Error message and console error: 

![image](https://user-images.githubusercontent.com/2676022/81947800-32fd3000-95be-11ea-8dc4-9cebb88049b7.png)

**Comment by adamsilverstein** [comment]
On the ga/provisioning branch when I click cancel then go to analytics I get these errors:

![image](https://user-images.githubusercontent.com/2676022/81948584-21685800-95bf-11ea-98be-a831675a1c29.png)

**Comment by adamsilverstein** [comment]
When I don't enable the provisioning scope:

![image](https://user-images.githubusercontent.com/2676022/81949256-de5ab480-95bf-11ea-9948-2e9af2873074.png)


I get this error back on the analytics screen:

![image](https://user-images.githubusercontent.com/2676022/81949159-c3884000-95bf-11ea-90cb-c6d2443c59b7.png)

When I click submit the request errors, however no error message is displayed. Fixing this.

**Comment by adamsilverstein** [comment]
Adding the error component helps a little, but the error is cryptic: 

![image](https://user-images.githubusercontent.com/2676022/81950878-bc623180-95c1-11ea-91ab-440ffa5edc99.png)

**Comment by felixarntz** [comment]
@cole10up The problems you mentioned regarding not granting scopes are general problems not introduced by this one, so they shouldn't block this. Could you create a separate issue with your findings there and move to "Prioritization"? This should apply to any time where you don't grant a certain permission in the OAuth flow.

I believe the other concerns were addressed by Evan's and Adam's comments, they should be resolved if you test the latest code in https://github.com/google/site-kit-wp/pull/1343.

**Comment by felixarntz** [comment]
Back to QA via #1343 (`develop` branch)

**Comment by cole10up** [comment]
**Tested**

Setup latest SK, activated Analytics with a gmail account that doesn't have an Analytics account.
![image](https://user-images.githubusercontent.com/57115907/82403238-36a41300-9a13-11ea-82de-8bbd1aca4335.png)

Clicked Cancel on the initial page.

Confirmed drives to Settings properly
![image](https://user-images.githubusercontent.com/57115907/82403278-4de30080-9a13-11ea-877b-0a00342f8ac8.png)

Clicking back into setup module functions properly and repopulates Analytics info

Adjusted the form and removed a character on each form and clicked 'Create Account' 
![image](https://user-images.githubusercontent.com/57115907/82405837-d95f9000-9a19-11ea-9cc7-9a74e5c65ad4.png)

Notice User is driven to Analytics in the same window and errored below;
![image](https://user-images.githubusercontent.com/57115907/82405745-9f8e8980-9a19-11ea-9c0e-d76c6afa0a14.png)

Clicking back button and trying again without adjusting forms notice same error occurred after clicking 'Create Account'
![image](https://user-images.githubusercontent.com/57115907/82405931-14fa5a00-9a1a-11ea-9343-aaea1bf1c3ab.png)

Tried one more time and was able to see TOS page
![Site Kit by Google Settings ‹ Hellogoodbye — WordPress (1)](https://user-images.githubusercontent.com/57115907/82406099-96ea8300-9a1a-11ea-95f8-207e557ebfd1.gif)

Hit back button and tried again - notice token Error.
![image](https://user-images.githubusercontent.com/57115907/82406151-b41f5180-9a1a-11ea-817c-81577514cee4.png)

This seems unstable and I feel needs some attention

**Comment by adamsilverstein** [comment]
Thanks for testing @cole10up - I will take a look.

**Comment by felixarntz** [comment]
@aaemnnosttv Is #1581 up for review? Still a draft :)

**Comment by aaemnnosttv** [comment]
@felixarntz It was basically ready for review but I've touched it up a bit and added a fix for something else in there too. It's ready for your review now, so I'm assigning this back to you 👍

**Comment by adamsilverstein** [comment]
Related PR which fixes a loading flash of the "zero data" component for Analytics if you click the left hand Analytics sub menu and Analytics is not yet configured: https://github.com/google/site-kit-wp/pull/1583

**Comment by aaemnnosttv** [comment]
@cole10up this one is ready for another pass!

Regarding errors that were previously seen on TOS screen - one case is expected if the user is not logged in with their Google account in the browser that they have connected to their Site Kit user. This is fairly unlikely to happen, as most users will probably have just activated the module and gone through OAuth right before. This is part of what we're working on in #1566 right now, so they will actually need to do this first because the provisioning scope won't be requested initially. All of the interface flickers should be cleaned up now as well.

Let me know if there's anything else that comes up in your testing here.

**Comment by cole10up** [comment]
**Tested**

Uploaded and activated latest SK with a user who doesn't have an analytics account. Ran through the setup of the module:
![image](https://user-images.githubusercontent.com/57115907/83095947-4c3fbb00-a059-11ea-8ce1-9c5a11689219.png)

Clicked create account

Notice receiving scopes error:
![image](https://user-images.githubusercontent.com/57115907/83097337-b7d75780-a05c-11ea-9f47-f73e609d620a.png)

See recording here: All scopes were approved.

![Sign in - Google Accounts (1)](https://user-images.githubusercontent.com/57115907/83097391-d2a9cc00-a05c-11ea-8706-ff27a0f331e8.gif)

Sending this back as this to me seems to be an issue somewhere in the scopes process.

Site: https://hellogoodbye.today/wp-admin/admin.php

**Comment by aaemnnosttv** [comment]
@cole10up This is expected as #1582 was just merged which is the first part of work that we're doing to change which scopes are initially requested, and to request more sensitive scopes (such as account provisioning) on-demand. This is not working now because the client side bit #1585 is not merged yet, but will be before the release. I'll give this another look over now but it should be good to go to Approval, as we have dedicated issues to QA regarding these newer changes.

**Comment by aaemnnosttv** [comment]
## QA

Testing with 1219d66de6c66ec97ced7be7f10d327716f72422

* ✅ Display new screen with all necessary fields (as many as possible pre-filled) for the user to approve or edit
    * ✅ Account: pre-filled name (Site Title) --> user should be able to edit this 
    * ✅ Property: pre-filled domain of the site URL --> user should be able to edit this 
    * ✅ ~~Profile~~ View: pre-filled (all website data) --> user should be able to edit this 
    * ✅ Timezone: pre-filled (from WP site's timezone) --> user should be able to pick this from dropdown 
        - Implemented as Country select + Timezone select
* ✅ Redirect smoothly to GA TOS for the user to approve (see https://developers.google.com/analytics/devguides/config/provisioning/v3/user-controlled#technical-tos-flow-url), using the account ticket ID generated via #1212
* ✅ Redirect to main plugin dashboard and show success message for the new account + it will take some time for data to aggregate (top banner) 
    - This is shown in an iframe on the Analytics service - nothing we can do about that right now
* ✅ Handle errors during the process and display relevant messages for user to take action 
* ✅ Finally take the user to the main dashboard with the standard congrats banner at the top.

More QA will happen around this feature in #1582 and #1585 regarding newer changes here. Moving to Approval.
